automata/fa/dfa.py

```diff
@@ -30,6 +30,8 @@ def __init__(self, *, states, input_symbols, transitions,
         self.initial_state = initial_state
         self.final_states = final_states.copy()
         self.allow_partial = allow_partial
+        self._word_cache = []
```

Comment:
@Tagl You should probably rebase on top of my latest commits to the `immutability` branch. I've made the DFA class immutable by overriding `__setattr__` and `__delattr__`—this looks to be using the old `DFA.__init__` code.

I would ask that you keep the `_word_cache` scoped to the methods that need it, passing it around as a function argument as needed.

Reply:
@Tagl That's a fair point. And the fact that this is a private (and volatile, i.e. caching) property means the fact that it changes isn't such a big deal.

Then my recommendation would be: when you rebase off the latest `immutability` branch, feel free to create the cache initially using the `super().__init__` call, then because you should not need to reassign to `_word_cache`, you should be able to push to it freely.

This will probably break my existing `test_dfa_immutable_dict` test because in that test, I am assuming that all attrs on the DFA's `__dict__` are hashable. But I can probably adjust that, or maybe expose a separate method for retrieving the `__dict__` with internal properties excluded.

@eliotwrobson What do you think?

Reply:
@caleb531 I think that course of action sounds reasonable. @Tagl Is right that these really should be stored along with the lifetime of the object, since having to recompute them is expensive, and it really should be done as-needed (the way that it is done here). I think testing for immutability excluding internal properties is a good plan.

Reply:
@eliotwrobson Agreed. I've also added a new `Automaton.attributes()` method to this commit that basically returns `__dict__` but with only the public attributes. I might rename this later to be more semantically-clear, but not until after this PR is merged.